tools/ci/run_tc.py

```diff
@@ -262,13 +266,60 @@ def setup_environment(args):
 
+def setup_repository(args):
+    is_pr = os.environ.get("GITHUB_PULL_REQUEST", "false") != "false"
+
+    # Initially task_head points at the same commit as the ref we want to test
+    # However that may not be the same commit as we actually want to test if
+    # the branch changed since the decision task ran.
+    # In that case we take a different approach depending on whether this is a
+    # PR or a push to a branch.
+    # If this is a push to a branch, and the original commit is still fetchable
+    # we try to fetch that (it may not be in the case of e.g. a force push).
+    # If it's not fetchable we fail the run
+    # For a PR we are testing the provisional merge commit. If that's changed it
+    # could be that the PR branch is updated or the base branch is updated. In the
+    # former case we fail the run because testing an old commit is just a waste of
+    # resources. In the latter case we assume it's OK to use the current merge
+    # instead of the one at the time the decision task ran.
+
+    if args.ref:
+        if is_pr:
+            assert args.ref.endswith("/merge")
+            expected_head = args.merge_rev
+        else:
+            expected_head = args.head_rev
+
+        task_head = run(["git", "rev-parse", "task_head"], return_stdout=True).strip()
+
+        if task_head != expected_head:
+            if not is_pr:
+                try:
+                    run(["git", "fetch", "origin", "%s:task_head" % expected_head])
+                except subprocess.CalledProcessError:
+                    print("CRITICAL: task_head points at %s, expected %s and "
+                          "unable to fetch expected commit.\n"
+                          "This may be because the branch was updated" % (task_head, args.rev))
+                    sys.exit(1)
+            else:
+                head_ref = args.ref.rsplit("/", 1)[0] + "/head"
+                try:
+                    remote_head = run(["git", "ls-remote", "origin", head_ref],
+                                      return_stdout=True).split("\t")[0]
+                except subprocess.CalledProcessError:
+                    print("CRITICAL: Failed to read remote ref %s" % head_ref)
+                    sys.exit(1)
+                if remote_head != args.head_rev:
+                    print("CRITICAL: task_head points at %s, expected %s. "
+                          "This may be because the branch was updated" % (task_head, args.rev))
+                    sys.exit(1)
+                print("INFO: Merge commit changed from %s to %s due to base branch changes. "
+                      "Running task anyway." % (expected_head, task_head))
 
-def setup_repository():
     if os.environ.get("GITHUB_PULL_REQUEST", "false") != "false":
-        parents = run(["git", "show", "--no-patch", "--format=%P", "task_head"], return_stdout=True).strip().split()
+        parents = run(["git", "show", "--no-patch", "--format=%P", "task_head"],
```

Comment:
Personally, I'd prefer using git porcelain commands in scripts, but this is fine, too.

`git rev-parse task_head^@`

Reply:
Sorry, I meant plumbing commands.